### PR TITLE
PCHR:3462 - Add Tool Icon to Edit Task Options

### PR DESCRIPTION
**Pull request by ajesamson**
## Overview
Editing of Task Type, Workflow and Document task options was not initially possible when creating records involving them without having to first navigate away. This PR add a wrench tool icon to each of the required pages in order to make editing of options more convenient and easily locate the editing page. 

## Before
before_task

before_document

before_workflow

## After
after_task

after_document

after_workflow

test

## Technical Details
Tool icon html was added to each of the affected pages. The wrench icon contain an ng-click call to page controller function, triggering opening of the dialogue window for editing (with the exception of workflow). 

```html
<a class="pointer">
    <i class="crm-i fa-wrench" ng-click="task.openActivityTypeOptionsEditor()"></i>
</a>
```

The openActivityTypeOptionsEditor function is called when user click on wrench icon, opening the option window for editing (using the crmAngService and passing the link option editing page). The options list is updated on closure of the window.

```javascript
crmAngService.loadForm('/civicrm/admin/options/activity_type?reset=1')
  .on('crmUnload', function () {
    taskService.getOptions().then(function (options) {
        angular.extend($rootScope.cache, options);
    });
  });
```

The task and document api usually return name instead of label. The name should not be editable but label should. This implies that changes made on the records will not reflect from API response. This was changed to ensure users get the true reflection of their activities.
```php
foreach ($types as $type) {
  if (!$sequential) {
    $result[$type['value']] = $type['label'];
  }
  else {
    $result[] = array(
      'key' => $type['value'],
      'value' => $type['label'],
    );
  }
}
```

In order to ensure the code base was not broken, new records were created for Task and Document. Both, reflected the changed label as soon as the dialogue window was closed.
 
## Comments
_Workflow page uses angular route which [CiviCRM currently have issues with](https://civicrm.stackexchange.com/questions/19817/how-can-i-put-an-angularjs-page-inside-a-crm-popup-modal-window) when opened in a dialogue box. Due to this, the page is opened in a new tab in order not to block the ticket_

---
✅Manual Tests - passed
✅Jasmine Tests - passed
✅JS distributive files - included

### CODE REVIEW

**Comment by igorpavlov** [comment]
@ajesamson please fix the indentation in your technical snippets.

**Comment by igorpavlov** [comment]
@ajesamson Such changes should also be covered in tech details section https://github.com/compucorp/civihr-tasks-assignments/pull/360/files#diff-ef27ca463094d4b48b92c65c2e61460cL656

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js [hunk 1]
```diff
@@ -597,6 +598,18 @@ define([
...
       $rootScope.$watch('cache.contact.arrSearch', collectContacts);
     }
+
+    /**
+     * Opens editor for activity type options editing
+     */
+    function openActivityTypeOptionsEditor () {
+      crmAngService.loadForm('/civicrm/admin/options/activity_type?reset=1')
+        .on('crmUnload', function () {
+          documentService.getOptions().then(function (options) {
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-document.controller.js]:
```js
documentService.getOptions()
  .then(function (options) {
    angular.extend($rootScope.cache, options);
  });
```
according to the styleguide https://compucorp.atlassian.net/wiki/spaces/PCHR/pages/121371880/AngularJS

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-task.controller.js [hunk 2]
```diff
@@ -258,6 +259,18 @@ define([
 
       return true;
     }
+
+    /**
+     * Opens editor for activity type options editing
+     */
+    function openActivityTypeOptionsEditor () {
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-task.controller.js]: This is a duplicate code with https://github.com/compucorp/civihr-tasks-assignments/pull/360/files#diff-76ee16aaa0874056d972e19b2f08802fR605, did you have any thoughts on how to avoid that?

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/task.service.js [hunk 3]
```diff
@@ -3,7 +3,8 @@
 define([
   'common/angular',
   'common/lodash',
-  'common/moment'
+  'common/moment',
+  'common/services/crm-ang.service'
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/task.service.js]: It does not seem like you use this service in that file, any reason why the injection is here?

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.controller.spec.js [hunk 4]
```diff
@@ -612,6 +614,31 @@ define([
       });
     });
 
+    describe('when user click on the document wrench icon', function () {
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.controller.spec.js]: when user click**s** on the document...

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.controller.spec.js [hunk 5]
```diff
@@ -612,6 +614,31 @@ define([
...
+      var onPopupFormSuccess;
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.controller.spec.js]: `var onPopupFormSuccess, url;`
you can merge vars together if there is no assignment

File: uk.co.compucorp.civicrm.tasksassignments/views/contact/tasks.html [hunk 6]
```diff
@@ -11,14 +11,17 @@
...
-        <div class="col-xs-12 col-sm-6 col-lg-4" ng-if="settings.extEnabled.assignments">
+        <div class="col-xs-11 col-sm-5 col-lg-3" ng-if="settings.extEnabled.assignments">
           <ui-select theme="civihr-ui-select" prevent-animations multiple ng-model="list.filterParams.assignmentType">
             <ui-select-match prevent-animations class="ui-select-match" placeholder="Select workflow...">{{$item.title}}</ui-select-match>
             <ui-select-choices prevent-animations class="ui-select-choices" repeat="type.id as type in cache.assignmentType.arr">
               <div ng-bind-html="type.title | highlight: $select.search"></div>
             </ui-select-choices>
           </ui-select>
         </div>
+        <div class="col-xs-1 col-sm-1 col-lg-1">
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/views/contact/tasks.html]: Please try to add a class `control-label` to align the wrench like here:

image

Also, I don't think you need col-sm-1 and col-lg-1.

```html
<div class="col-xs-1 control-label">
```

File: uk.co.compucorp.civicrm.tasksassignments/views/contact/documents.html [hunk 7]
```diff
@@ -16,14 +16,17 @@
...
-              <div class="col-xs-12 col-sm-9">
+              <div class="col-xs-11 col-sm-8">
                 <ui-select theme="civihr-ui-select" prevent-animations multiple ng-model="list.filterParams.assignmentType">
                   <ui-select-match prevent-animations class="ui-select-match" placeholder="Select workflow...">{{$item.title}}</ui-select-match>
                   <ui-select-choices prevent-animations class="ui-select-choices" repeat="type.id as type in cache.assignmentType.arr">
                     <div ng-bind-html="type.title | highlight: $select.search"></div>
                   </ui-select-choices>
                 </ui-select>
               </div>
+              <div class="col-xs-1 col-sm-1">
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/views/contact/documents.html]: I don't think you need "col-sm-1" class here, try to get rid of it and see if it works.

File: uk.co.compucorp.civicrm.tasksassignments/views/modal/assignment.html [hunk 8]
```diff
@@ -29,14 +29,17 @@ <h4 class="modal-title">New Workflow</h4>
...
-          <div class="col-xs-12 col-sm-6">
+          <div class="col-xs-11 col-sm-6">
             <div class="crm_custom-select crm_custom-select--full">
               <select name="assignment" ng-class="{'has-error': assignmentForm.assignment.$invalid}" ng-required="true" class="form-control no-select2" ng-change="setData()" ng-model="assignment.case_type_id" ng-options="value.id as value.title for value in cache.assignmentType.arr">
                 <option value="">- select -</option>
               </select>
               <span class="crm_custom-select__arrow"></span>
             </div>
           </div>
+          <div class="col-xs-1 col-sm-1">
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/views/modal/assignment.html]: Unnecessary class, same as above

File: uk.co.compucorp.civicrm.tasksassignments/views/modal/document.html [hunk 9]
```diff
@@ -56,7 +56,12 @@ <h4 class="modal-title" ng-switch="documentModal.isRole('admin')">
       </ui-select>
     </div>
   </div>
-</div>
+    <div class="col-xs-1">
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/views/modal/document.html]: Please add a class to align, same as above.

File: uk.co.compucorp.civicrm.tasksassignments/views/modal/task.html [hunk 10]
```diff
@@ -4,7 +4,7 @@
       <span aria-hidden="true">&times;</span>
       <span class="sr-only">Close</span>
     </button>
-    <h4 class="modal-title">New Task</h4>
+    <h4 class="modal-title">New Tasks</h4>
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/views/modal/task.html]: Why new Tasks? Can you now create more than one task at a time?

File: uk.co.compucorp.civicrm.tasksassignments/views/modal/task.html [hunk 11]
```diff
@@ -34,15 +34,26 @@ <h4 class="modal-title">New Task</h4>
...
+                    <ui-select-match prevent-animations allow-clear class="ui-select-match" placeholder="Select Task Type">
+                      {{$select.selected.value}}
+                    </ui-select-match>
+                    <ui-select-choices prevent-animations class="ui-select-choices" repeat="type.key as type in cache.taskType.arr | filter: $select.search">
+                      <div ng-bind="type.value"></div>
+                    </ui-select-choices>
+                  </ui-select>
+                </div>
+              </div>
+              <div class="col-xs-2">
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/views/modal/task.html]: Please add a class to align, same as above.

**Comment by igorpavlov** [comment]
@ajesamson Tasks components are used in SSP as well. This is why you have combos of `col-xs` `col-sm` `col-lg` etc - because SSP is mobile responsive. This PR lacks GIFs for SSP where you should show how it responds to viewport dimensions change.

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-task.controller.js (hunk 2 above)
Comment (ajesamson) [reply uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-task.controller.js]: Not really as i think the two of them belongs to different scope.

File: uk.co.compucorp.civicrm.tasksassignments/api/v3/Task.php [hunk 12]
```diff
@@ -601,12 +601,12 @@ function civicrm_api3_task_getoptions($params) {
...
       if (!$sequential) {
-        $result[$type['value']] = $type['name'];
+        $result[$type['value']] = $type['label'];
       }
       else {
         $result[] = array(
           'key' => $type['value'],
-          'value' => $type['name'],
+          'value' => $type['label'],
         );
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.tasksassignments/api/v3/Task.php]: @ajesamson please mention on the ticket description why these changes to the Task and Document APIs were necessary and what parts of the system did you test to make sure this didn't break anything

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/task.service.js (hunk 3 above)
Comment (ajesamson) [reply uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/task.service.js]: It does actually unless i am adding it in a wrong place. Without it, the `loadForm` is not recognized and my tests fail.

File: uk.co.compucorp.civicrm.tasksassignments/api/v3/Task.php (hunk 12 above)
Comment (ajesamson) [reply uk.co.compucorp.civicrm.tasksassignments/api/v3/Task.php]: I have the ticket description. Kindly check again please.

File: uk.co.compucorp.civicrm.tasksassignments/views/modal/document.html (hunk 9 above)
Comment (ajesamson) [reply uk.co.compucorp.civicrm.tasksassignments/views/modal/document.html]: Can you expatiate on this. Should i move the tool icon closer to the drop menu?

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-task.controller.js (hunk 2 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/controllers/modal/modal-task.controller.js]: It's fine actually, I looked at it with a fresh eye and despite of the fact the functions look the same and do the same I don't think we should move them to a separate service of something, that may be an overkill.

File: uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/task.service.js (hunk 3 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/services/task.service.js]: This looks wrong to me. I think the service dependency should be here actually:

*uk.co.compucorp.civicrm.tasksassignments/js/src/tasks-assignments/modules/tasks-assignments.services.js* cc @AkA84

File: uk.co.compucorp.civicrm.tasksassignments/views/modal/document.html (hunk 9 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.tasksassignments/views/modal/document.html]: I meant the `control-label` class that aligns it vertically.

File: uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.controller.spec.js [hunk 13]
```diff
@@ -614,9 +614,8 @@ define([
       });
     });
 
-    describe('when user click on the document wrench icon', function () {
-      var onPopupFormSuccess;
-      var url;
+    describe('when user clicks on the document wrench icon', function () {
+      var onPopupFormSuccess, url;
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.tasksassignments/js/test/controllers/modal/modal-document.controller.spec.js]: Now I can see that you may simply want to move these vars on 1 *describe* level higher to about duplication of var definition. Also, if you could improve the name of the var `url` it would be great, I would suggest `popupFormUrl`

**Comment by igorpavlov** [comment]
Please add screenshots/gifs of SSP, just to make sure all looks right, even if the wrench is not there.
